tensorflow_quantum/python/optimizers/rotosolve_minimizer.py

```diff
@@ -0,0 +1,342 @@
+    ...        loss_value = hinge_loss(model(x_circ, training=True), Y)
+    ...        return loss_value
+
+    Last you can run the minimize algorithm. The initial parameter is
+    guessed randomly.
+
+    >>> rotosolve_minimizer.minimize(model_func,np.random.rand([2]))
+
+    Args:
+        expectation_value_function:  A Python callable that accepts
+            a point as a real `tf.Tensor` and returns a `tf.Tensor`s
+            of real dtype containing the value of the function.
+            The function to be minimized. The input is of shape `[n]`,
+            where `n` is the size of the trainable parameters.
+            The return value is a real `tf.Tensor` Scalar (matching shape
+            `[1]`).  This must be a linear combination of quantum
+            measurement expectation value, otherwise this algorithm cannot
+            work.
+        initial_position: Real `tf.Tensor` of shape `[n]`. The starting
+            point, or points when using batching dimensions, of the search
+            procedure. At these points the function value and the gradient
+            norm should be finite.
+        tolerance: Scalar `tf.Tensor` of real dtype. Specifies the tolerance
+            for the procedure. If the supremum norm between two iteration
+            vector is below this number, the algorithm is stopped.
+        name: (Optional) Python `str`. The name prefixed to the ops created
+            by this function. If not supplied, the default name 'minimize'
+            is used.
+
+    Returns:
+        optimizer_results: A RotosolveOptimizerResults object contains the
+            result of the optimization process.
+    """
+
+    with tf.name_scope(name or 'minimize'):
+        initial_position = tf.convert_to_tensor(initial_position,
+                                                name='initial_position',
+                                                dtype='float32')
+        dtype = initial_position.dtype.base_dtype
+        tolerance = tf.convert_to_tensor(tolerance,
+                                         dtype=dtype,
+                                         name='grad_tolerance')
+        max_iterations = tf.convert_to_tensor(max_iterations,
+                                              name='max_iterations')
+
+        def _rotosolve_one_parameter_once(state):
+            """Rotosolve a single parameter once.
+
+            Args:
+                state: A RotosolveOptimizerResults object stores the
+                       current state of the minimizer.
+
+            Returns:
+                states: A list which the first element is the new state
+            """
+            delta_shift = tf.reshape(
+                tf.cast(tf.sparse.to_dense(
+                    tf.sparse.SparseTensor(
+                        [[state.solve_param_i, 0]], [math.pi / 2],
+                        [prefer_static_shape(state.position)[0], 1])),
+                        dtype=dtype), prefer_static_shape(state.position))

```

Comment:
Would it be possible to use `scatter_update` or `scatter_nd` here instead of these sparse -> dense + casts + reshapes ?

Reply:
Fixed in 8aee2e6bc4ab3bfdd1edb41f184fb669e023d01f